cmake/common.cmake

```diff
@@ -45,3 +45,29 @@ function(HIDE_DISTRACTING_VARIABLES)
   endif()
 endfunction()
 
+function(GROUP_SOURCE_TO_DIR_STRUCTURE)
+  if(MSVC)
+    foreach(_F ${ARGN})
+      string(REGEX MATCH "${CMAKE_CURRENT_SOURCE_DIR}" _M ${_F})
+      # Basic test for determining if current file is a source file or header file.
+      if(${_F} MATCHES ".+\.cpp$")
+        set(_F_PREFIX "Source Files\\")
+      else()
+        set(_F_PREFIX "Header Files\\")
+      endif()
+      # Get relative path from current source directory or current binary directory.
+      if(${_F} MATCHES "^${CMAKE_CURRENT_SOURCE_DIR}.+")
+        file(RELATIVE_PATH _L "${CMAKE_CURRENT_SOURCE_DIR}" ${_F})
+      else()
+        file(RELATIVE_PATH _L "${CMAKE_CURRENT_BINARY_DIR}" ${_F})
+        # Assume all files in the current binary directory are generated header files!
+        set(_F_PREFIX "Generated Header Files\\")
+      endif()
+      get_filename_component(_G ${_L} DIRECTORY)
+      if(_G)
+        string(REPLACE "/" "\\" _G ${_G})
+      endif()
+      source_group("${_F_PREFIX}${_G}" FILES "${_F}")
+    endforeach()
+  endif()
+endfunction()
```

Comment:
I don't think using FILE would be appropriate, CMake might have something to say about that!  I'll change the variable names to something 'proper' but not dangerous.